sub/dec_sub.c

```diff
@@ -193,6 +195,30 @@ struct dec_sub *sub_create(struct mpv_global *global, struct track *track,
     return NULL;
 }
 
+// Called locked.
+static void update_cached_packets(struct dec_sub *sub, struct demux_packet *pkt,
+                                  double video_pts)
+{
+    // Unconditionally cleared.
+    if (sub->cached_pkts[PKT_CACHE - 1])
+        TA_FREEP(&sub->cached_pkts[PKT_CACHE - 1]);
+
+    // Always save at least 2 packets but allow up to PKT_CACHE
+    // for subs with overlapping durations.
+    int i = PKT_CACHE;
+    while (--i) {
+        struct demux_packet *prev_pkt = sub->cached_pkts[i - 1];
+        if (prev_pkt) {
+            double prev_pkt_end = prev_pkt->pts + prev_pkt->duration;
+            if (sub->cached_pkts[i] && prev_pkt_end < video_pts)
+                TA_FREEP(&sub->cached_pkts[i]);
+            sub->cached_pkts[i] = prev_pkt;
+            sub->cached_pkts[i - 1] = NULL;
```

Comment:
I'm probably not understanding it correctly, but shouldn't that go in an else branch?

If `sub->cached_pkts[i] && prev_pkt_end >= video_pts` then `sub->cached_pkts[i]` doesn't get freed but it still gets overwritten by `sub->cached_pkts[i - 1]`.

Reply:
But then it's not guaranteed that `sub->cached_pkts[0]` is free for another packet...

Reply:
>it still gets overwritten by sub->cached_pkts[i - 1].

That's was intended to always advance the cache but indeed it never frees anything until the cache maxes out.